From 7a8622e7f3e06cd2714cf50485646513cf77423b Mon Sep 17 00:00:00 2001
From: zachary <zhangzg@marvell.com>
Date: Wed, 28 Dec 2016 16:38:51 +0800
Subject: [PATCH 1220/2241] fix: spi: a3700: limit output clock up to max spi
 clock frequency

The SPI slave devices can work with a higher clock frequency than
the maximum output clock (100MHz) supported by Armada3700. Each SPI
transfer will pass the speed_hz to the SPI driver. In case the
devices requires a higher clock rate, SPI driver always applies the
maximum output clock to the device.

Change-Id: Ifbb87ea53f30ef9c947a5e68d29fdbee7d1ac0cc
Signed-off-by: zachary <zhangzg@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/34803
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Tested-by: Wilson Ding <dingwei@marvell.com>
---
 drivers/spi/spi-armada-3700.c | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/spi/spi-armada-3700.c b/drivers/spi/spi-armada-3700.c
index 8f4323c..be24509 100644
--- a/drivers/spi/spi-armada-3700.c
+++ b/drivers/spi/spi-armada-3700.c
@@ -263,14 +263,20 @@ static void a3700_spi_mode_set(struct a3700_spi *a3700_spi,
 	spireg_write(a3700_spi, A3700_SPI_IF_CFG_REG, val);
 }
 
-static int a3700_spi_baudrate_set(struct a3700_spi *a3700_spi,
+static int a3700_spi_clock_set(struct a3700_spi *a3700_spi,
 	unsigned int speed_hz)
 {
 	u32 val;
 	u32 prescale;
 
-	/* calculate Prescaler = (spi_input_freq / spi_max_freq) */
-	prescale = a3700_spi->input_clk_freq / a3700_spi->max_clk_freq;
+	/*
+	* compare SPI control input frequency and flash input frequency
+	* to choose lower frequency for using
+	*/
+	if (speed_hz >= a3700_spi->max_clk_freq)
+		prescale = a3700_spi->input_clk_freq / a3700_spi->max_clk_freq;
+	else
+		prescale = DIV_ROUND_UP(a3700_spi->input_clk_freq, speed_hz);
 
 	val = spireg_read(a3700_spi, A3700_SPI_IF_CFG_REG);
 	val = val & ~A3700_SPI_CLK_PRESCALE_MASK;
@@ -476,9 +482,9 @@ static int a3700_spi_transfer_setup(struct spi_device *spi,
 
 	if (!(a3700_spi->flags & AUTO_CS) && !a3700_spi->status.cs_active) {
 		/* Set SPI clock prescaler */
-		ret = a3700_spi_baudrate_set(a3700_spi, xfer->speed_hz);
+		ret = a3700_spi_clock_set(a3700_spi, xfer->speed_hz);
 		if (ret) {
-			dev_err(&spi->dev, "set baudrate failed\n");
+			dev_err(&spi->dev, "set clock failed\n");
 			goto out;
 		}
 
-- 
2.7.4

